metricbeat/module/kafka/partition/partition.go

```diff
@@ -0,0 +1,103 @@
+package partition
+
+import (
+	"github.com/Shopify/sarama"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/metricbeat/mb"
+)
+
+// init registers the partition MetricSet with the central registry.
+func init() {
+	if err := mb.Registry.AddMetricSet("kafka", "partition", New); err != nil {
+		panic(err)
+	}
+}
+
+// MetricSet type defines all fields of the partition MetricSet
+type MetricSet struct {
+	mb.BaseMetricSet
+}
+
+// New create a new instance of the partition MetricSet
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+
+	config := struct{}{}
+
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+
+	return &MetricSet{
+		BaseMetricSet: base,
+	}, nil
+}
+
+// Fetch partition stats list from kafka
+func (m *MetricSet) Fetch() ([]common.MapStr, error) {
+
+	config := sarama.NewConfig()
+	config.Net.DialTimeout = m.Module().Config().Timeout
+	config.Net.ReadTimeout = m.Module().Config().Timeout
+	config.ClientID = "metricbeat"
+
+	client, err := sarama.NewClient([]string{m.Host()}, config)
+	if err != nil {
+		return nil, err
+	}
+	defer client.Close()
+
+	topics, err := client.Topics()
+	if err != nil {
+		return nil, err
+	}
+
+	events := []common.MapStr{}
+	for _, topic := range topics {
+		partitions, err := client.Partitions(topic)
+		if err != nil {
+			logp.Err("Fetch partition info for topic %s: %s", topic, err)
+		}
+
+		for _, partition := range partitions {
+			newestOffset, err := client.GetOffset(topic, partition, sarama.OffsetNewest)
+			if err != nil {
+				logp.Err("Fetching newest offset information for partition %s in topic %s: %s", partition, topic, err)
+			}
+
+			oldestOffset, err := client.GetOffset(topic, partition, sarama.OffsetOldest)
+			if err != nil {
+				logp.Err("Fetching oldest offset information for partition %s in topic %s: %s", partition, topic, err)
+			}
+
+			// List of replicas ids for the given partition
+			replicas, err := client.Replicas(topic, partition)
+			if err != nil {
+				logp.Err("Fetch replicas for partition %s in topic %s: %s", partition, topic, err)
+			}
+
+			brocker, err := client.Leader(topic, partition)
+			if err != nil {
+				logp.Err("Fetching brocker for partition %s in topic %s: %s", partition, topic, err)
+			}
+
+			event := common.MapStr{
+				"topic":     topic,
+				"partition": partition,
+				"offset": common.MapStr{
+					"oldest": oldestOffset,
+					"newest": newestOffset,
+				},
+				"replicas": replicas,
+				"broker": common.MapStr{
+					"id":      brocker.ID(),
+					"address": brocker.Addr(),
+				},
+			}
```

Comment:
Hmm... I was in hope for getting the offset per replica and create an event per replicate. The `broker` is the current leader to a partition. All other brokers handling the same partition will be the replica. Every now and then the leader is re-elected, turning the broker into an replica and another former replica into the leader.

Some event like:

```
common.MapStr {
  "topic": topic,
  "partition": partition,
  "broker": common.MapStr {
    "id": id,
    "address": brokers[id],
  }
  "offset": offset,
  "leader": id == leaderID,
}
```

The address of every broker and broker id is included in meta-data.